src/func.hpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2019 IBM Corp. and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef WINTER_FUNC_HPP
+#define WINTER_FUNC_HPP
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "environment.hpp"
+#include "type.hpp"
+#include "wassert.hpp"
+
+namespace winter {
+
+class ModuleInstance;
+class ModuleInstanceInternal;
+
+struct UnlinkedFuncInternal;
+class UnlinkedFunc;
+
+struct LinkedFuncInternal;
+class LinkedFunc;
+
+using JitFunction = uint32_t (*)(LinkedFuncInternal* func);
+
+class InstructionStream {
+    std::vector<uint8_t> _stream;
+public:
+    InstructionStream(std::vector<uint8_t>&& stream) : _stream(stream) {}
+
+    size_t size() const { return _stream.size(); }
+
+    friend class InstructionCursor;
+};
+
+class InstructionCursor {
+    const InstructionStream* _stream;
+    const uint8_t* _cursor;
```

Comment:
My original thinking was that `InstructionCursor` could be standard layout so that we could use it for our call stack. However, in designing things, I've realized that's not such a good idea.

That being said, is there a particular reason _to_ use `std::vector<uint8_t>::iterator` here instead? I don't really see the benefit of doing that...

Reply:
I guess there isn't really any benefit. `std::vector<uint8_t>::iterator` _should_ just be `uint8_t*` anyways so using the first over the latter would be functionally identical. It's more a question of style, really. I'm fine with using `uint8_t*` then.